src/server/modules/auth/SgidLoginController.ts

```diff
@@ -0,0 +1,89 @@
+import Express, { CookieOptions } from 'express'
+import { inject, injectable } from 'inversify'
+import { SGID_LOGIN_OAUTH_STATE, SgidAuthService } from '../../services/sgid'
+import { UserRepositoryInterface } from '../../repositories/interfaces/UserRepositoryInterface'
+import { DependencyIds } from '../../constants'
+import { validEmailDomainGlobExpression } from '../../config'
+
+export const OFFICER_EMAIL_SCOPE = 'ogpofficerinfo.work_email'
+const SGID_STATE_COOKIE_NAME = 'gogovsg_sgid_state'
+const SgidStateCookieConfig: CookieOptions = {
+  httpOnly: true,
+}
+
+@injectable()
+export class SgidLoginController {
+  private sgidService
+
+  private userRepository: UserRepositoryInterface
+
+  constructor(
+    @inject(DependencyIds.userRepository)
+    userRepository: UserRepositoryInterface,
+  ) {
+    this.sgidService = SgidAuthService
+    this.userRepository = userRepository
+  }
+
+  public generateAuthUrl: (
+    req: Express.Request,
+    res: Express.Response,
+  ) => void = (_req, res) => {
+    const { url, codeVerifier, nonce } = this.sgidService.authorizationUrl()
+
+    res.cookie(
+      SGID_STATE_COOKIE_NAME,
+      { codeVerifier, nonce },
+      SgidStateCookieConfig,
+    )
+    res.send(url)
+    return
+  }
+
+  public handleLogin = async (req: Express.Request, res: Express.Response) => {
+    try {
+      const { code, state } = req.query
+      const sessionData = req.cookies[SGID_STATE_COOKIE_NAME]
+
+      if (state !== SGID_LOGIN_OAUTH_STATE) {
+        throw new Error('Error logging in with SGID: state is incorrect')
```

Comment:
what might you think about handling this similarly to line 70, i.e. something like `res.redirect('/#/ogp-login?statusCode=400'); return`, which at least sends the user back to the frontend where they can get some indication that something went wrong?

this failure mode with sgid state also seems to indicate more of a client fault with a bad request; meanwhile, throwing an `Error` probably gives an internal server error (?) indicating a server fault, which might not be our intention

Reply:
Hmm makes sense! We can do that!